Fixed continuous_domain to account for nested fractions

ThomasHickman opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #18857
Fixes #17018
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
I've refined the logic in continuous_domain to search out for nested fraction (i.e. `log(1/x)`) to count them as singularities. sec, csc, cot and tan are also expanded out in their fractional form, in terms of cos to enable this to be done better.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* calculus
  * fixed continuous_domain to account for nested fractions
<!-- END RELEASE NOTES -->

ThomasHickman committed:
Fix #18857 (bug in continuous_domain)

smichr commented on sympy/calculus/util.py:
```diff
@@ -85,19 +86,11 @@ def continuous_domain(f, symbol, domain):
-                    solveset(denom(together(f)), symbol, domain)
+        sings = EmptySet()
+        for atom in f.rewrite([sec, csc, cot, tan], cos).atoms(Pow):
+            base, exponent = atom.args
```
or `base, exponent = atom.as_base_exp()`

smichr replied:
I suspect you should simply put this rewriting at the top of the `try` that you deleted and allow it to work as it did before.

ThomasHickman replied:
I've replaced this with `as_base_exp` now (and done all the squashing and testing). The old code would not work with the addition of this rewriting, as it: firstly doesn't look at fractions which are nested inside other expressions (as in sometime like `sin(1/x)` and secondly using `solveset` on `1/f(x)` for an `f(x)` that isn't a fraction will never result in a solution, due to `solveset` not working in any situation where it has to do algebra with infinites. This issue is reported at #18857, but it doesn't look like that's going to be resolved anytime soon, so I thought I would use this alternative method

smichr replied:
Hmm...I agree with you. What was there for finding singularities does not look like it would work well.

It looks like there are two steps: find the domain and then look for singularities in the domain. I'm not sure why `Abs` was targetted since that is a "absolute continuity" idea. And it's not clear from the docstring which is intended. And no tests are failing without that code. e.g. `sqrt(x - abs(x -3))` will have a domain limited to non-negative args for the `sqrt`; in addition there will be a break in differentiability (but not continuity) at `x = 3`.

smichr replied:
Looking at other functions in the file, however, I see `function_range`. The only thing that would differentiate these two functions is if the `continuous_domain` considered differentiability in which case it should be taking `Abs` into account.

ThomasHickman replied:
I think `function_range` generates the range of a function, not it's domain? Also, from the PR of the previous logic: (#13335 and notably https://github.com/sympy/sympy/pull/13335#discussion_r140632892), it looks the different cases were introduced to work around https://github.com/sympy/sympy/issues/12527. `Intersection(solveset(1/f, symbol), domain)` is preferable to as it returns the correct answer, but if an expression contains an `Abs`, solveset will not work with a complex domain, and if you have a power of 2 a infinite recursion happens (you can see this by trying the test with `Intersection(solveset(1/f, symbol), domain)` everywhere with the old code).
Anyway, I don't think we'll have to use these cases in the new logic as #13335 isn't a problem as we're not going to be dealing with infinities which can be eliminated later on.

ThomasHickman replied:
Also, I'm assuming that `continuous_domain` is dealing with "non-absolute continuity", as this is the current behaviour. As you've said, it may be better if the documentation is clear up whether this is the case or not

smichr replied:
> generates the range of a function, not it's domain?

doh - myopia

smichr commented:
If you test this locally in an interactive session `test('test_util')` you will see that it doesn't pass the existing tests: there are 4 failures. When it passes those locally, then push your changes here, otherwise Travis gets backed up doing this simple step and your commit history keeps growing. Locally you can make commits and when your have everything working you can squash the changes into a single commit.  It just helps keep everything working and looking better.

smichr commented:
Looking good. Note, you could also just refer to the `exp` and `base` attributes of a Pow as in `if atom.exp.is_negative`...but it's not important to make that change.

oscarbenjamin commented:
I tried testing this PR and it doesn't seem to work in isympy:
```julia
In [1]: from sympy.calculus.util import continuous_domain                                                                                                                         

In [2]: continuous_domain(log(tan(x)**2+1), x, S.Reals)                                                                                                                           
Out[2]: (-∞, ∞)
```

smichr commented:
> I tried testing this PR and it doesn't seem to work in isympy:

I don't use isympy but when I try to do so in one case I get what you get but then under another scenario I see
```python
>>> from sympy.calculus.util import continuous_domain
>>> continuous_domain(log(tan(x)**2+1), x, S.Reals)
cos(x)**(-2) cos(x) True
cos(x - pi/2)**2 cos(x - pi/2) False
          ⎛⎧        π        ⎫   ⎧        3⋅π        ⎫⎞
(-∞, ∞) \ ⎜⎨2⋅n⋅π + ─ | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬⎟
          ⎝⎩        2        ⎭   ⎩         2         ⎭⎠
````
which makes me think it is a printing issue.

smichr commented:
OK, I got ipython installed and when I run the test in master I get (-oo,oo) and when I run it under this branch I get what I printed above. I think this is ready to go but will wait until we hit 24 hours in PR.

oscarbenjamin commented:
It's not a printing issue. The test code fails:
```julia
In [1]: from sympy.calculus.util import continuous_domain                                                                                                                         

In [2]:     domain = continuous_domain(log(tan(x)**2+1), x, S.Reals) 
   ...:     assert not domain.contains(3/2*pi) 
   ...:     assert domain.contains(5)                                                                                                                                             
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-b98a2f66fc7e> in <module>
      1 domain = continuous_domain(log(tan(x)**2+1), x, S.Reals)
----> 2 assert not domain.contains(3/2*pi)
      3 assert domain.contains(5)

AssertionError: 
```
I think that `3/2*pi` should be `S(3)/2*pi` to avoid floats but that isn't the problem since it actually gives an interval:
```
In [3]: continuous_domain(log(tan(x)**2+1), x, S.Reals)                                                                                                                           
Out[3]: (-∞, ∞)
```
There is a bug to do with caching. I tested in the debugger and the code added here is not reached at all under isympy.

oscarbenjamin commented:
Okay it seems to have been a problem with my setup. I reinstalled the git checkout of sympy in the virtual env and it works now.

I do think that the test should not use floats though:
```
In [12]: 3/2*pi                                                                                                                                                                   
Out[12]: 1.5⋅π
```

smichr commented on sympy/calculus/tests/test_util.py:
```diff
@@ -67,6 +67,9 @@ def test_continuous_domain():
     assert continuous_domain(1/(x**2 - 4) + 2, x, S.Reals) == \
         Union(Interval.open(-oo, -2), Interval.open(-2, 2), Interval.open(2, oo))
+    domain = continuous_domain(log(tan(x)**2+1), x, S.Reals)
+    assert not domain.contains(3/2*pi)
```
```suggestion
    assert not domain.contains(3*pi/2)
```

smichr commented on sympy/calculus/tests/test_util.py:
```diff
@@ -67,6 +67,9 @@ def test_continuous_domain():
         Union(Interval.open(-oo, 0), Interval.open(0, oo))
     assert continuous_domain(1/(x**2 - 4) + 2, x, S.Reals) == \
         Union(Interval.open(-oo, -2), Interval.open(-2, 2), Interval.open(2, oo))
+    domain = continuous_domain(log(tan(x)**2+1), x, S.Reals)
```
```suggestion
    domain = continuous_domain(log(tan(x)**2 + 1), x, S.Reals)
```

smichr committed:
enhanced singularities

smichr commented:
My `cd` branch improves singularity handling and calls that function from `continuous_domain`. Could you cherry pick the last commit to your branch?

ThomasHickman commented:
Just changed @smichr commit to correct the docstring (it previously said that a NotImplementedError was thrown for an input of irrational functions). Also, this fixes #17018, so I've added this to the PR body

ThomasHickman committed:
Fix set_function to work with the new singularities

ThomasHickman commented:
^ just fixed things here to make `_set_function` in `sympy/sets/handlers/functions.py` work, as it calls `singularities`

smichr commented on sympy/sets/handlers/functions.py:
```diff
@@ -66,6 +66,10 @@ def _set_function(f, x): # noqa:F811
     if not x.start.is_comparable or not x.end.is_comparable:
         return
 
+    # XXX: when non-rational functions are supported, remove this line. 
```
```suggestion
    # XXX: when non-rational functions are supported, remove this line.
```

smichr replied:
If this is needed then the trailing space must be removed. (There is a setting for git that can be used to do this automatically, btw.)

But why is this needed?

ThomasHickman replied:
When `singularities` was not updated, this function would just return None for non-rational functions, as `singularities` would throw a `NotImplementedError`. I've just replicated the same behaviour as making this function support non-rational functions may take some non-trivial effort and I think this is best done in another PR (for a start, `singularities` needs to be replaced by `continuous_domain` or even some function like `function_domain`)

smichr replied:
I pushed a commit to my `cd` branch that relaxes this constraint and fixes issue #12560

smichr committed:
let singularities raise; fix #12560

ThomasHickman commented:
I've cherrypicked @smichr commit and added:
```python
        sing = singularities(expr, var, x)

        if not (isinstance(sing, FiniteSet) or sing is EmptySet):
            raise NotImplementedError()
```
as sing could be an infinite list (for example when dealing with `imageset(x, 1/cos(x), S.Reals)`). I've also added `imageset(x, 1/cos(x), S.Reals)` as a test case, asserting that this doesn't hang.

smichr commented on sympy/sets/handlers/functions.py:
```diff
@@ -82,7 +89,10 @@ def _set_function(f, x): # noqa:F811
         _end = f(x.end)
 
     if len(sing) == 0:
-        solns = list(solveset(diff(expr, var), var))
```
Can this change be reverted now?

smichr replied:
What test fails without this? All `sets` tests pass for me.

ThomasHickman replied:
This is still necessary, if the differential of `expr` solves to an infinite expression, but doesn't have an infinite amount of singularities. The tests don't actually test this at the moment, but if you apply the patch that you've put, `imageset(x, cos(x), S.Reals)` will hang. I'll add this as another test case now
